com.chartboost.helium/Android~/helium-android-bridge/src/main/kotlin/com/chartboost/heliumsdk/unity/HeliumUnityBridge.kt

```diff
@@ -0,0 +1,390 @@
+    private var rewardedEventListener: IRewardedEventListener? = null
+    private var ilrdObserver: HeliumIlrdObserver? = null
+    fun setupEventListeners(
+        lifeCycleListener: ILifeCycleEventListener?,
+        interstitialListener: IInterstitialEventListener?,
+        rewardedListener: IRewardedEventListener?,
+        bannerListener: IBannerEventListener?
+    ) {
+        lifeCycleEventListener = lifeCycleListener
+        interstitialEventsListener = interstitialListener
+        rewardedEventListener = rewardedListener
+        bannerEventsListener = bannerListener
+    }
+
+    fun setSubjectToCoppa(isSubject: Boolean) {
+        HeliumSdk.setSubjectToCoppa(isSubject)
+    }
+
+    fun setSubjectToGDPR(isSubject: Boolean) {
+        HeliumSdk.setSubjectToGDPR(isSubject)
+    }
+
+    fun setCCPAConsent(hasGivenConsent: Boolean) {
+        HeliumSdk.setCCPAConsent(hasGivenConsent)
+    }
+
+    fun setUserHasGivenConsent(hasGivenConsent: Boolean) {
+        HeliumSdk.setUserHasGivenConsent(hasGivenConsent)
+    }
+
+    var userIdentifier: String?
+        get() = HeliumSdk.getUserIdentifier()
+        set(userIdentifier) {
+            HeliumSdk.setUserIdentifier(userIdentifier)
+        }
+
+    fun onBackPressed(): Boolean {
+        return HeliumSdk.onBackPressed()
+    }
+
+    fun start(appId: String?, appSignature: String?, unityVersion: String?) {
+        _activity = UnityPlayer.currentActivity
+
+        ilrdObserver = object : HeliumIlrdObserver {
+            override fun onImpression(impData: HeliumImpressionData) {
+                val json = serializePlacementILRDData(impData.placementId, impData.ilrdInfo)
+                lifeCycleEventListener!!.DidReceiveILRD(json)
+            }
+        }
+        runTaskOnUiThread {
+            // This call initializes the Helium SDK. This might change in the future with two ID parameters and we'll get rid of the logControllerListener
+            HeliumSdk.start(_activity, appId!!, appSignature!!) { error: Error? ->
+                val errorNotFound = error == null
+                if (errorNotFound) {
+                    Log.d("Unity", "HeliumUnityBridge: Plugin Initialized")
+                    HeliumSdk.setGameEngine("unity", unityVersion)
+                    HeliumSdk.subscribeIlrd(ilrdObserver!!)
+                } else {
+                    Log.d("Unity", "HeliumUnityBridge: Plugin failed to initialize: $error")
+                }
+                val errorCode = if (errorNotFound) -1 else 1
+                val errorDescription = if (errorNotFound) EMPTY_STRING else error.toString()
+                lifeCycleEventListener!!.DidStart(errorCode, errorDescription)
+            }
+        }
+    }
+
+    fun setTestMode(mode: Boolean) {
+        HeliumSdk.setTestMode(mode)
+    }
+
+    fun destroy() {
+        if (ilrdObserver != null) {
+            HeliumSdk.unsubscribeIlrd(ilrdObserver!!)
```

Comment:
you don't want to unsub null! you should check

Reply:
isn't it already checked in the if statement before the unsubscribe call? Or is there any other way to do it? @bichenwang

Reply:
ah! ilrdObserver is a member variable! It could have changed between the check and the call. imagine a multithreaded environment where one thread is setting `ilrdObserver` to the actual observer and another thread is setting `ilrdObserver` to `null` as fast as they can. There's a good chance that `ilrdObserver` will not be null on the first check but by the time it's accessed again, it's null! so the `ilrdObserver!!` call crashes since it's now null.
two ways to fix this:
This is the kotlin idiomatic way:
```
ilrdObserver?.let {
    HeliumSdk.unsubscribeIlrd(it)
}
```
The other way which is more useful when you're checking a bunch of null things (this assigns a local variable that doesn't change even if the underlying variable does change):
```
val observer = ilrdObserver
if (observer != null) {
    HeliumSdk.unsubscribeIlrd(observer)
}
```

Reply:
Thanksss this is hugeee for understanding it now :) much appreciated.